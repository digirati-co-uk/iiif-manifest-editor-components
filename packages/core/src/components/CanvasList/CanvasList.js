import React from 'react';
import PropTypes from 'prop-types';
import { Droppable, Draggable } from 'react-beautiful-dnd';
import { IconButton, withStyles } from '@material-ui/core';
import { Cancel, AddCircle } from '@material-ui/icons';
import Panel from '../Panel/Panel';
import LocaleString from '../LocaleString/LocaleString';
import Tooltip from '../DefaultTooltip/DefaultTooltip';

const grid = 8;

const getItemStyle = (isDragging, draggableStyle, isSelected) => ({
  // change background colour if dragging
  background: isDragging ? 'rgb(89, 191, 236)' : 'rgb(236, 231, 231)',
  color: isDragging ? 'rgb(255,255,255)' : 'rgb(0, 0, 0)',
  outline: isSelected ? '2px solid rgb(89, 191, 236)' : '0',

  // styles we need to apply on draggables
  ...draggableStyle,
});

const getListStyle = (isDraggingOver, draggableStyle) => ({
  background: isDraggingOver ? 'white' : 'white',
  //...draggableStyle,
});

const style = theme => ({
  droppable: {
    position: 'relative',
    display: 'flex',
    padding: theme.spacing.unit,
    overflow: 'auto',
    minHeight: '100%',
  },
  listItem: {
    userSelect: 'none',
    padding: `${theme.spacing.unit / 2}px ${theme.spacing.unit * 2}px`,
    margin: `0 ${theme.spacing.unit}px 0 0`,
    display: 'flex',
    alignItems: 'flex-end',
    maxWidth: 180,
    position: 'relative',
    height: 100,
    boxSizing: 'border-box',
  },
  canvas: {
    whiteSpace: 'nowrap',
    textOverflow: 'ellipsis',
    overflow: 'hidden',
  },
  deleteButton: {
    position: 'absolute',
    top: -10,
    right: -10,
  },
  defaultAddButtonSpacer: {
    display: 'flex',
    alignItems: 'center',
  },
  noCanvasesContainer: {
    position: 'absolute',
    top: 0,
    left: 0,
    bottom: 0,
    right: 0,
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center',
  },
});

const emptyFn = () => {};

const CanvasList = ({
  classes,
  children,
  canvases,
  toolbar,
  selected,
  lang,
  select,
  remove,
  invokeAction,
}) => (
  <Panel horizontal={true}>
    {toolbar && <Panel.Toolbar>{toolbar}</Panel.Toolbar>}
    <Panel.Content>
      <Droppable droppableId="canvaslist" direction="horizontal">
        {(providedDroppable, snapshotProppable) => (
          <div
            ref={providedDroppable.innerRef}
            style={getListStyle(snapshotProppable.isDraggingOver)}
            className={classes.droppable}
            {...providedDroppable.droppableProps}
          >
            {canvases && canvases.length > 0 ? (
              <React.Fragment>
                {canvases.map((canvas, index) => (
                  <Draggable
                    key={canvas.id}
                    draggableId={canvas.id}
                    index={index}
                  >
                    {(provided, snapshot) => (
                      <div
                        ref={provided.innerRef}
                        {...provided.draggableProps}
                        {...provided.dragHandleProps}
                        style={getItemStyle(
                          snapshot.isDragging,
                          provided.draggableProps.style,
                          selected === canvas.id
                        )}
                        className={classes.listItem}
                      >
                        {typeof children === 'function' ? (
                          children(canvas, remove, select)
                        ) : (
                          <div
                            key={`canvas_list_${canvas.id}`}
                            className={classes.canvas}
                            title={
                              canvas.label && canvas.label[lang]
                                ? canvas.label[lang]
                                : canvas.id
                            }
                            onClick={() => select(canvas)}
                          >
                            <LocaleString fallback={canvas.id} lang={lang}>
                              {canvas.label}
                            </LocaleString>
                          </div>
                        )}
                        <Tooltip title="Delete Canvas">
                          <IconButton
                            onClick={() => remove(canvas)}
                            className={classes.deleteButton}
                          >
                            <Cancel />
                          </IconButton>
                        </Tooltip>
                      </div>
                    )}
                  </Draggable>
                ))}
                <div className={classes.defaultAddButtonSpacer}>
                  <Tooltip title="Add Canvas">
                    <IconButton onClick={() => invokeAction('add-canvas')}>
                      <AddCircle />
                    </IconButton>
                  </Tooltip>
                </div>
              </React.Fragment>
            ) : (
              <div className={classes.noCanvasesContainer}>
                No canvases in the manifest,
                <Tooltip title="Add">
                  <IconButton onClick={() => invokeAction('add-canvas')}>
                    <AddCircle />
                  </IconButton>
                </Tooltip>
                a canvas.
              </div>
            )}
            {providedDroppable.placeholder}
          </div>
        )}
      </Droppable>
    </Panel.Content>
  </Panel>
);

CanvasList.propTypes = {
  /* JSS classes */
  classes: PropTypes.any,
  /* if a function passed, the rendered of the children can be overridden */
  children: PropTypes.any,
  /* list of canvases to display */
  canvases: PropTypes.array,
  /* custom toolbar */
  toolbar: PropTypes.any,
  /* list of the selected canvases */
  selected: PropTypes.string,
  /* current language  */
  lang: PropTypes.string,
  /* on annotation seleced callback */
  select: PropTypes.func,
  /* on remove callback */
  remove: PropTypes.func,
  /* toolbar action dispacher */
  invokeAction: PropTypes.func.isRequired,
};

CanvasList.defaultProps = {
  selected: null,
  select: emptyFn,
  remove: emptyFn,
  invokeAction: emptyFn,
};

export default withStyles(style)(CanvasList);
