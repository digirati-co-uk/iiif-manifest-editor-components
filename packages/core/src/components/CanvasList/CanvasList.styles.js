export default theme => ({
  droppable: {
    position: 'relative',
    display: 'flex',
    //padding: theme.spacing.unit,
    padding: 0,
    overflow: 'auto',
    minHeight: '100%',
  },
  droppableVertical: {
    position: 'relative',
    padding: theme.spacing.unit,
    overflow: 'auto',
    maxWidth: '100%',
  },
  listItem: {
    userSelect: 'none',
    margin: `${theme.spacing.unit}px ${theme.spacing.unit / 2}px`,
    display: 'flex',
    alignItems: 'stretch',
    maxWidth: 110,
    minWidth: 110,
    position: 'relative',
    height: 100,
    boxSizing: 'border-box',
  },
  listItemDragging: {
    background: theme.palette.primary.main,
    color: theme.palette.secondary.main,
  },
  listItemSelected: {
    outline: `2px solid ${theme.palette.primary.main}`,
  },
  listItemVertical: {
    userSelect: 'none',
    margin: `0 ${theme.spacing.unit}px 0 0`,
    display: 'inline-block',
    alignItems: 'stretch',
    width: `calc(50% - ${theme.spacing.unit}px)`,
    minWidth: 100,
    position: 'relative',
    height: 100,
    boxSizing: 'border-box',
  },
  canvas: {
    position: 'relative',
    height: '100%',
    width: '100%',
    textAlign: 'center',
  },
  canvasThumbnail: {
    maxWidth: 150,
    haxHeight: 100,
    height: '100%',
  },
  canvasLabel: {
    whiteSpace: 'nowrap',
    textOverflow: 'ellipsis',
    overflow: 'hidden',
    position: 'absolute',
    bottom: 0,
    right: 0,
    left: 0,
    zIndex: 1,
    color: 'white',
    textAlign: 'center',
    backgroundColor: 'rgba(0,0,0,0.2)',
    padding: `${theme.spacing.unit / 2}px ${theme.spacing.unit * 2}px`,
  },
  deleteButton: {
    position: 'absolute',
    top: -10,
    right: -10,
  },
  defaultAddButtonSpacer: {
    display: 'flex',
    alignItems: 'center',
  },
  noCanvasesContainer: {
    position: 'absolute',
    top: 0,
    left: 0,
    bottom: 0,
    right: 0,
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center',
  },
});
