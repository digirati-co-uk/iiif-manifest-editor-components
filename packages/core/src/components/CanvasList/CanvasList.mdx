---
name: Canvas List
menu: 02. High Level Components
---

import { Playground, PropsTable } from 'docz';
import CanvasList from './CanvasList'; 

# Canvas List

Component that lists canvases in the specified manifest.

## Requirements

- [X] add drag and drop list using the same drag drop manager as all the other components
- [x] add selectable state for canvases
- [x] use fesk-bem for themeability
- [x] use the global selected resources
- [x] make sure that it is able to work as an idividual component
- [x] being able to override the default annotation renderer
- [x] create default toolbar component
- [x] add delete button for the individual canvases

<PropsTable of={CanvasList} />

