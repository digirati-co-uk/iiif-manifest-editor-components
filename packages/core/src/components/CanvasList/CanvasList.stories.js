import * as React from 'react';
import { storiesOf } from '@storybook/react';
import { withKnobs, text } from '@storybook/addon-knobs';

import CanvasList from './CanvasList.js';
import { DragDropContext } from 'react-beautiful-dnd';

storiesOf('CanvasList', module)
  .addDecorator(withKnobs)
  .add('CanvasList', () => {
    const TEST_CANVASES = [
      {
        id: 'test-canvas-1',
        type: 'Canvas',
      },
      {
        id: 'test-canvas-2',
        type: 'Canvas',
      },
    ];
    return (
      <DragDropContext onDragEnd={() => {}}>
        <CanvasList
          direction={text('Horizontal', 'horizontal')}
          canvases={TEST_CANVASES}
          selected={text('Selected', 'test-canvas-1')}
        />
      </DragDropContext>
    );
  });
