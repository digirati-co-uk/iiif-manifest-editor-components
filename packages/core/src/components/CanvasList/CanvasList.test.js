import React from 'react';
import CanvasList from './CanvasList';

describe('CanvasList', () => {
  it('renders the canvas list unchanged', () => {
    const TEST_CANVASES = [
      {
        id: 'test-canvas-1',
      },
      {
        id: 'test-canvas-2',
      },
    ];

    const canvasList = shallow(<CanvasList canvases={TEST_CANVASES} />);
    expect(canvasList).toMatchSnapshot();
  });
});
