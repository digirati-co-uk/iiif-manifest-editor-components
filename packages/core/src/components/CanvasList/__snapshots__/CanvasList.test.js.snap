// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CanvasList renders the canvas list unchanged 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <WithStyles(CanvasList)
    canvases={
      Array [
        Object {
          "id": "test-canvas-1",
        },
        Object {
          "id": "test-canvas-2",
        },
      ]
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "canvases": Array [
        Object {
          "id": "test-canvas-1",
        },
        Object {
          "id": "test-canvas-2",
        },
      ],
      "classes": Object {
        "canvas": "CanvasList-canvas-7",
        "canvasLabel": "CanvasList-canvasLabel-9",
        "canvasThumbnail": "CanvasList-canvasThumbnail-8",
        "defaultAddButtonSpacer": "CanvasList-defaultAddButtonSpacer-11",
        "deleteButton": "CanvasList-deleteButton-10",
        "droppable": "CanvasList-droppable-1",
        "droppableVertical": "CanvasList-droppableVertical-2",
        "listItem": "CanvasList-listItem-3",
        "listItemDragging": "CanvasList-listItemDragging-4",
        "listItemSelected": "CanvasList-listItemSelected-5",
        "listItemVertical": "CanvasList-listItemVertical-6",
        "noCanvasesContainer": "CanvasList-noCanvasesContainer-12",
      },
      "direction": "horizontal",
      "invokeAction": [Function],
      "itemClass": null,
      "listClass": null,
      "remove": [Function],
      "select": [Function],
      "selected": null,
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "canvases": Array [
          Object {
            "id": "test-canvas-1",
          },
          Object {
            "id": "test-canvas-2",
          },
        ],
        "classes": Object {
          "canvas": "CanvasList-canvas-7",
          "canvasLabel": "CanvasList-canvasLabel-9",
          "canvasThumbnail": "CanvasList-canvasThumbnail-8",
          "defaultAddButtonSpacer": "CanvasList-defaultAddButtonSpacer-11",
          "deleteButton": "CanvasList-deleteButton-10",
          "droppable": "CanvasList-droppable-1",
          "droppableVertical": "CanvasList-droppableVertical-2",
          "listItem": "CanvasList-listItem-3",
          "listItemDragging": "CanvasList-listItemDragging-4",
          "listItemSelected": "CanvasList-listItemSelected-5",
          "listItemVertical": "CanvasList-listItemVertical-6",
          "noCanvasesContainer": "CanvasList-noCanvasesContainer-12",
        },
        "direction": "horizontal",
        "invokeAction": [Function],
        "itemClass": null,
        "listClass": null,
        "remove": [Function],
        "select": [Function],
        "selected": null,
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
